core/services/ping/ping1d_mavlink.py

```diff
@@ -0,0 +1,154 @@
+
+    ## Send distance_sensor message to autopilot
+    async def send_distance_data(self, distance: int, deviceid: int, confidence: int) -> None:
+        # logger.debug("sending distance %d confidence %d" % (distance, confidence))
+        if confidence < 0.5:
+            distance = 0
+        logger.debug(f"sendind {distance}")
+        await self.mavlink2rest.send_mavlink_message(
+            self.distance_message(int((time.time() - self.tboot) * 1000), int(distance / 10), deviceid)
+        )
+
+    ## Send a request for distance_simple message to ping device
+    async def send_ping1d_request(
+        self,
+    ) -> None:
+        logger.debug("requesting new data")
+        data = PingMessage()
+        data.request_id = PING1D_DISTANCE_SIMPLE
+        data.src_device_id = 0
+        data.pack_msg_data()
+        self.ping1d_io.sendall(data.msg_data)
+
+    # pylint: disable=too-many-branches
+    async def drive(self, port: int) -> None:
+
+        ## Parser to decode incoming PingMessage
+        ping_parser = PingParser()
+
+        ## Messages that have the current distance measurement in the payload
+        distance_messages = [PING1D_DISTANCE, PING1D_DISTANCE_SIMPLE, PING1D_PROFILE]
+
+        ## The last time a distance measurement was received
+        last_distance_measurement_time = 0.0
+
+        ## The last time a distance measurement was requested
+        last_ping_request_time = 0.0
+
+        pingserver = ("127.0.0.1", port)
+        self.ping1d_io.connect(pingserver)
+
+        # set the ping interval once at startup
+        # the ping interval may change if another client to the pingproxy requests it
+        data = PingMessage()
+        data.request_id = PING1D_SET_PING_INTERVAL
+        data.src_device_id = 0
+        data.ping_interval = int(PING_INTERVAL_MS * 1000)
+        data.pack_msg_data()
+        self.ping1d_io.sendall(data.msg_data)
+
+        while True:
+            if not self.should_run:
+                await asyncio.sleep(1)
+                continue
+            await asyncio.sleep(0.001)
+            tnow = time.time()
+
+            # request data from ping device
+            if tnow > last_distance_measurement_time + PING_INTERVAL_MS:
+                if tnow > last_ping_request_time + PING_INTERVAL_MS:
+                    last_ping_request_time = time.time()
+                    await self.send_ping1d_request()
+
+            # deal with possibly lost connection
+            if tnow > last_distance_measurement_time + PING_INTERVAL_MS * 10:
+                logger.info("attempting reconnection...")
+                self.ping1d_io.connect(pingserver)
+                await asyncio.sleep(0.1)
+
+            # read data in from ping device
+            try:
+                data, _ = self.ping1d_io.recvfrom(4096)
+            except socket.error as exception:
+                # check if it's waiting for data
+                if exception.errno == errno.EAGAIN:
+                    continue
```

Comment:
true. Though a wise man once said "explicit is better than impicit". I do like having EAGAIN there so we know it happens and is being taking into consideration...

Reply:
It could be in a comment, above the general `continue` for that loop?
Not a huge deal, I'm just not a fan of multiple branches that do exactly the same thing :-)

Reply:
I'm inclined to agree with @ES-Alexander here, why `EAGAIN` is special over others that needs his special `continue` ?

Reply:
after thinking about this for a while, I realized the corrrect approach is to use `select` to check if there's anything to receive, instead of doing it blindly